test/decl/func/dynamic_self.swift

```diff
@@ -330,3 +330,61 @@ class Runce : Runcible {
     return self
   }
 }
+
+// ----------------------------------------------------------------------------
+// Forming a type with 'Self' in invariant position
+
+struct Generic<T> { init(_: T) {} }
+
+class InvariantSelf {
+  func me() -> Self {
+    let a = Generic(self)
+    let _: Generic<InvariantSelf> = a
+    // expected-error@-1 {{cannot convert value of type 'Generic<Self>' to specified type 'Generic<InvariantSelf>'}}
+
+    return self
+  }
+}
+
+// FIXME: This should be allowed
+
+final class FinalInvariantSelf {
+  func me() -> Self {
+    let a = Generic(self)
+    let _: Generic<FinalInvariantSelf> = a
+    // expected-error@-1 {{cannot convert value of type 'Generic<Self>' to specified type 'Generic<FinalInvariantSelf>'}}
+
+    return self
+  }
+}
+
+// ----------------------------------------------------------------------------
+// Semi-bogus factory init pattern
+
+protocol FactoryPattern {
+  init(factory: @autoclosure () -> Self)
+}
+
+extension  FactoryPattern {
+  init(factory: @autoclosure () -> Self) { self = factory() }
+}
+
+class Factory : FactoryPattern {
+  init(_string: String) {}
+
+  convenience init(string: String) {
+    self.init(factory: Factory(_string: string))
+    // expected-error@-1 {{incorrect argument label in call (have 'factory:', expected '_string:')}}
+    // FIXME: Bogus diagnostic
```

Comment:
This seems rather sub par to remove an acceptable workaround for a deficiency in the language. Are framework developers expected just not have compatibility to objective-c? Given this we are effectively either re-introducing bugs and preventing proper pairity with Foundation or are limiting swift-corelibs-foundation to be stuck on swift 4. 

I have a whole branch that refractors almost ALL of the classes that follow the self replacement pattern. TBH that is most of the classes in Foundation. I would prefer to have a suitable replacement in place before pulling the rug out from under developers using a feature (even if it is "semi-bogus" or "unsound" as you put it).